core/src/main/scala/kafka/server/BrokerServer.scala

```diff
@@ -397,6 +397,9 @@ class BrokerServer(
       info("shutting down")
 
       if (config.controlledShutdownEnable) {
+        // Shut down the broker metadata listener, so that we don't get added to any
+        // more ISRs.
+        brokerMetadataListener.beginShutdown()
```

Comment:
If we do this, does `brokerMetadataListener.close() `still need to call `beginShutdown()`.

Reply:
It does need to be because there are some paths through the code that don't go through here.  In general calling `beginShutdown` or `close` multiple times is harmless-- only the first time has an effect.